### PR TITLE
Generalize the idea of a "plan mode", vs just destroy flag

### PR DESCRIPTION
**Pull request by apparentlymart**
Previously there were only two planning modes: normal mode and destroy mode. In that context it made sense for these to be distinguished only by a boolean flag.

We're now getting ready to add our third mode, "refresh only". This establishes the idea that planning can be done in one of a number of mutually-exclusive "modes", which are related to but separate from the various other options that serve as modifiers for the plan operation.

This commit only introduces the new `plans.Mode` type and replaces the existing "destroy" flag fields with a field of that type. This doesn't cause any change in effective behavior because Terraform Core still supports only `NormalMode` and `DestroyMode`, with `NewContext` rejecting an attempt to create a `RefreshMode` context for now, and the CLI layer having no codepath that tries to set it.

It is in retrospect a little odd that the "destroy" flag was part of `ContextOpts` rather than just an argument to the `Plan` method, but refactoring that would be too invasive a change for right now so we'll leave this as a field of the context for now and save revisiting that for another day.

---

For this PR I've just peeled off the first two commits from #28297. These two just change internal implementation details of how we represent destroy mode without adding any new functionality. I want to land these first and early because they are cross-cutting and thus likely to come into conflict with other work if they were part of a long-lived feature branch.

With these in place the remaining work should involve more localized commits that we can land without necessarily colliding with other work in other subsystems.

I've marked this for backport primarily because otherwise it seems likely that other unrelated backports to v0.15 will end up failing due to this being a cross-cutting change. However, it also seems likely that at least parts of the planned functionality for refresh-only plans will also land in the v0.15 branch, and so this backport will support that.

### CODE REVIEW

File: command/arguments/apply.go
Comment (apparentlymart) [inline command/arguments/apply.go:107]: Note: technically this branch is showing up a little early because nothing else here supports `plans.RefreshOnlyMode`, but it's dead code for the moment since `ParseApply` has no codepath where it would set `PlanMode` to this value. I just wanted to add this function all in one step to reduce churn a little.

**Comment by apparentlymart** [comment]
Looking again at what I cherry-picked here I see that this also brings in the idea of `terraform apply -destroy` and states that `terraform destroy` is an alias for it, which _is_ different than before. It doesn't allow users to achieve anything they couldn't previously achieve, but it does add some new CLI surface area for something that was already possible a different way.

Given that, my inclination is to run with this as-is rather than to try to carve that second commit into two parts, but I'd be willing to do that carving if others have concerns about it.

File: command/plan.go
Comment (mildwonkey) [inline command/plan.go:198]: I like this reorganization!

**Review by mildwonkey — COMMENTED** [review]
LGTM, I don't think any more carving is necessary.

File: command/arguments/plan_test.go
Comment (alisdair) [inline command/arguments/plan_test.go:60]: Ah, this is much neater, thanks.

**Review by alisdair — APPROVED** [review]
This looks good to me! I have a PR coming soon which would conflict with the operation view changes here, so I'll wait until this is merged and rebase.

**Pull request merged**
